File: android/app/src/main/java/com/singword/app/data/remote/LrclibLyricsDataSource.kt
package com.singword.app.data.remote

import android.util.Log
import com.singword.app.BuildConfig
import java.io.IOException
import java.net.Proxy
import java.util.concurrent.TimeUnit
import okhttp3.OkHttpClient
import retrofit2.HttpException
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class LrclibLyricsDataSource : LyricsDataSource, LyricsCandidateDataSource {
    override val providerName: String = "lrclib"

    private val api: LrclibApi by lazy {
        val client = OkHttpClient.Builder()
            .proxy(Proxy.NO_PROXY)
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .addInterceptor { chain ->
                val request = chain.request().newBuilder()
                    .header("User-Agent", "SingWord/1.0.0")
                    .build()
                chain.proceed(request)
            }
            .build()

        Retrofit.Builder()
            .baseUrl("https://lrclib.net/")
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(LrclibApi::class.java)
    }

    override suspend fun search(query: String): LyricsResult {
        return when (val candidateResult = searchCandidates(query, limit = 1)) {
            is LyricsCandidateResult.Success -> {
                val selected = candidateResult.candidates.firstOrNull()
                if (selected == null) {
                    LyricsResult.NotFound
                } else {
                    LyricsResult.Success(
                        trackName = selected.trackName,
                        artistName = selected.artistName,
                        lyrics = selected.lyrics,
                        provider = selected.provider
                    )
                }
            }
            is LyricsCandidateResult.NotFound -> LyricsResult.NotFound
            is LyricsCandidateResult.NetworkError -> LyricsResult.NetworkError(candidateResult.message)
            is LyricsCandidateResult.ProviderError -> LyricsResult.ProviderError(
                provider = candidateResult.provider,
                message = candidateResult.message
            )
        }
    }

    override suspend fun searchCandidates(query: String, limit: Int): LyricsCandidateResult {
        return try {
            val tracks = api.search(query)
                .filter { !it.plainLyrics.isNullOrBlank() }
                .take(limit.coerceAtLeast(1))

            if (tracks.isEmpty()) {
                LyricsCandidateResult.NotFound
            } else {
                LyricsCandidateResult.Success(
                    candidates = tracks.map { track ->
                        LyricsCandidate(
                            trackName = track.trackName ?: query,
                            artistName = track.artistName.orEmpty(),
                            lyrics = track.plainLyrics.orEmpty(),
                            provider = providerName
                        )
                    },
                    provider = providerName
                )
            }
        } catch (e: IOException) {
            if (BuildConfig.DEBUG) {
                Log.e("SingWord/Lrclib", "Network failure while searching lrclib for \"$query\"", e)
            }
            LyricsCandidateResult.NetworkError("网络异常，请检查连接后重试")
        } catch (e: HttpException) {
            if (BuildConfig.DEBUG) {
                Log.e("SingWord/Lrclib", "HTTP ${e.code()} while searching lrclib for \"$query\"", e)
            }
            LyricsCandidateResult.ProviderError(providerName, "歌词服务异常（HTTP ${e.code()}）")
        } catch (e: Exception) {
            if (BuildConfig.DEBUG) {
                Log.e("SingWord/Lrclib", "Unexpected failure while searching lrclib for \"$query\"", e)
            }
            LyricsCandidateResult.ProviderError(providerName, e.message ?: "歌词服务异常")
        }
    }
}

File: ios/singword/Networking/LrclibLyricsDataSource.swift
import Foundation

private struct LrclibTrack: Decodable {
    let id: Int?
    let trackName: String?
    let artistName: String?
    let albumName: String?
    let duration: Double?
    let plainLyrics: String?
    let syncedLyrics: String?
}

final class LrclibLyricsDataSource: LyricsDataSource, LyricsCandidateDataSource {
    let providerName: String = "lrclib"

    private let session: URLSession

    init(session: URLSession? = nil) {
        if let session {
            self.session = session
        } else {
            let config = URLSessionConfiguration.default
            config.timeoutIntervalForRequest = 10
            config.timeoutIntervalForResource = 10
            self.session = URLSession(configuration: config)
        }
    }

    func search(query: String) async -> LyricsResult {
        switch await searchCandidates(query: query, limit: 1) {
        case .success(let candidates, _):
            guard let selected = candidates.first else {
                return .notFound
            }
            return .success(
                trackName: selected.trackName,
                artistName: selected.artistName,
                lyrics: selected.lyrics,
                provider: selected.provider
            )
        case .notFound:
            return .notFound
        case .networkError(let message):
            return .networkError(message)
        case .providerError(let provider, let message):
            return .providerError(provider: provider, message: message)
        }
    }

    func searchCandidates(query: String, limit: Int = 5) async -> LyricsCandidateResult {
        guard var components = URLComponents(string: "https://lrclib.net/api/search") else {
            return .providerError(provider: providerName, message: "无效的歌词服务地址")
        }
        components.queryItems = [
            URLQueryItem(name: "q", value: query)
        ]
        guard let url = components.url else {
            return .providerError(provider: providerName, message: "无效的请求参数")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("SingWord/1.0.0", forHTTPHeaderField: "User-Agent")

        do {
            let (data, response) = try await session.data(for: request)
            if let httpResponse = response as? HTTPURLResponse,
               !(200...299).contains(httpResponse.statusCode) {
                return .providerError(
                    provider: providerName,
                    message: "歌词服务异常（HTTP \(httpResponse.statusCode)）"
                )
            }

            let tracks = try JSONDecoder().decode([LrclibTrack].self, from: data)
            let candidates = tracks
                .filter { !($0.plainLyrics?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true) }
                .prefix(max(1, limit))
                .map { track in
                    let normalizedTrackName = track.trackName?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                    return LyricsCandidate(
                        trackName: normalizedTrackName.isEmpty ? query : normalizedTrackName,
                        artistName: track.artistName ?? "",
                        lyrics: track.plainLyrics ?? "",
                        provider: providerName,
                        duration: track.duration
                    )
                }

            if candidates.isEmpty {
                return .notFound
            }

            return .success(candidates: candidates, provider: providerName)
        } catch let error as URLError {
            _ = error
            return .networkError("网络异常，请检查连接后重试")
        } catch {
            return .providerError(provider: providerName, message: error.localizedDescription)
        }
    }
}
